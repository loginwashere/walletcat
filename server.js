const express = require('express');
const crypto = require('crypto');
const bodyParser = require('body-parser');
const expressJwt = require('express-jwt');
const jwt = require('jsonwebtoken');
const format = require('date-fns/format');

const hashSecret = 'abcdefg';
const jwtSecret = 'shhhhhhared-secret';

function hashPassword(password) {
  return crypto
    .createHmac('sha256', hashSecret)
    .update(password)
    .digest('hex')
}

function md5(string) {
  return crypto
    .createHash('md5')
    .update(string)
    .digest("hex");
}

let users = [];
users.push({
  id: 1,
  email: 'admin@mail.com',
  username: 'admin',
  avatar: `https://www.gravatar.com/avatar/${md5('admin@mail.com')}?s=50`,
  password: hashPassword('qwerty'),
  created: format(new Date()),
  updated: format(new Date())
});

let currencies = [];
currencies.push({
  id: 1,
  name: 'USD',
  description: 'United States Dollar',
  created: format(new Date()),
  updated: format(new Date())
});
currencies.push({
  id: 2,
  name: 'UAH',
  description: 'Ukrainian Hryvna',
  created: format(new Date()),
  updated: format(new Date())
});

let tokens = [];

let userCurrencies = [];
userCurrencies.push({
  id: 1,
  userId: users[0].id,
  currencyId: currencies[0].id,
  created: format(new Date()),
  updated: format(new Date())
});

let categories = [];
categories.push({
  id: 1,
  userId: users[0].id,
  name: 'Fast food',
  description: 'Some fast food',
  created: format(new Date()),
  updated: format(new Date())
})

let accounts = [];
accounts.push({
  id: 1,
  userId: users[0].id,
  currencyId: userCurrencies[0].id,
  name: 'Wallet',
  description: 'My Wallet',
  amount: 0,
  created: format(new Date()),
  updated: format(new Date())
});
accounts.push({
  id: 2,
  userId: users[0].id,
  currencyId: userCurrencies[0].id,
  name: 'Stash',
  description: 'My Stash',
  amount: 0,
  created: format(new Date()),
  updated: format(new Date())
});

let transactions = [];
transactions.push({
  id: 1,
  userId: users[0].id,
  fromAccountId: accounts[0].id,
  toAccountId: accounts[1].id,
  categoryId: categories[0].id,
  fromAmount: 100,
  toAmount: 100,
  description: 'description',
  date: format(new Date()),
  created: format(new Date()),
  updated: format(new Date())
});

const app = express();

const jwtMiddleware = expressJwt({
    secret: jwtSecret
  })
  .unless({
    path: [
      { url: '/' },
      { url: '/accounts' },
      { url: '/transactions' },
      { url: '/categories' },
      { url: '/recurring-payments' },
      { url: '/currencies' },
      { url: '/reports' },
      { url: '/about' },
      { url: '/contact-us' },
      { url: '/profile' },
      { url: '/logout' },
      { url: '/login' },
      { url: '/register' },
      { url: '/static' },
      {
        url: '/api/auth',
        methods: ['POST']
      },
      {
        url: '/api/users',
        methods: ['POST']
      }
    ]
  });

function errorHandler(err, req, res, next) {
  if (err) {
    console.log(err);
  }
  if (err.name === 'UnauthorizedError') {
    res.status(401).json({
      error: 'invalid token...'
    });
  }
}

app.use(bodyParser.json());
app.use(jwtMiddleware);
app.use(errorHandler);
// Express only serves static assets in production
if (process.env.NODE_ENV === 'production') {
  app.use(express.static('build'));
}

app.set('port', (process.env.PORT || 3001));

app.get('/api/currencies', (req, res) => {
  res.json({
    currencies
  });
});

app.post('/api/users', (req, res) => {
  if (!req.body.email || !req.body.password) {
    res.status(400).json({
      error: 'Validation error'
    })
  }
  const user = {
    id: users.length + 1,
    email: req.body.email,
    avatar: `https://www.gravatar.com/avatar/${md5(req.body.email)}?s=50`,
    username: req.body.username,
    password: hashPassword(req.body.password),
    created: format(new Date()),
    updated: format(new Date())
  };
  users.push(user);
  res.json({
    user
  })
});

function generateToken(userId) {
  return jwt.sign(
    { sub: userId },
    jwtSecret,
    { expiresIn: '1h' }
  );
}

function getValidToken(tokens, userId) {
  const token = tokens[userId];
  if (token) {
    try {
      jwt.verify(token, jwtSecret);
      return token;
    } catch (err) {
      if (err && err.name === 'TokenExpiredError') {
        return generateToken(userId);
      }
      throw(err);
    }
  }
  return generateToken(userId);
}

app.post('/api/auth', (req, res) => {
  const user = users.filter(user => user.email === req.body.email)[0];
  if (!user) {
    return res.status(404).json({
      error: 'User not found'
    });
  }
  if (!req.body.password || hashPassword(req.body.password) !== user.password) {
    return res.status(403).json({
      error: 'Credentials do not match'
    });
  }
  const token = getValidToken(tokens, user.id);
  if (tokens[user.id] !== token) {
    tokens[user.id] = token;
  }
  return res.json({
    token,
    user
  });
});

app.delete('/api/auth', (req, res) => {
  tokens[req.user.sub] = null;
  res.json({
    result: true
  });
});

app.get('/api/user-currencies', (req, res) => {
  res.json({
    userCurrencies: userCurrencies
      .filter((userCurrency) => userCurrency.userId === req.user.sub)
  });
});

app.post('/api/user-currencies', (req, res) => {
  const newUserCurrecny = {
    id: userCurrencies.length + 1,
    userId: req.user.sub,
    currencyId: req.body.currencyId,
    created: format(new Date()),
    updated: format(new Date())
  };
  userCurrencies.push(newUserCurrecny)
  res.json(newUserCurrecny);
});

app.delete('/api/user-currencies/:id', (req, res) => {
  const userCurrency = userCurrencies
    .filter((userCurrency) => userCurrency.userId === req.user.sub)
    .filter((userCurrency) => userCurrency.id === req.params.id)
    [0];
  if (userCurrency) {
    const index = userCurrencies.indexOf(userCurrency);
    userCurrencies.splice(index, 1);
  }
  res
    .status(204)
    .json();
});

app.get('/api/categories', (req, res) => {
  res.json({
    categories: categories
      .filter(category => category.userId === req.user.sub)
  });
});

app.post('/api/categories', (req, res) => {
  const category = {
    id: categories.length + 1,
    userId: req.user.sub,
    name: req.body.name,
    description: req.body.description,
    created: format(new Date()),
    updated: format(new Date())
  };
  categories.push(category)
  res.json(category);
});

app.delete('/api/categories/:id', (req, res) => {
  const category = categories
    .filter((category) => category.userId === req.user.sub)
    .filter((category) => category.id === req.params.id)
    [0];
  if (category) {
    const index = categories.indexOf(category);
    categories.splice(index, 1);
  }
  res
    .status(204)
    .json();
});

app.get('/api/accounts', (req, res) => {
  res.json({
    accounts: accounts
      .filter(account => account.userId === req.user.sub)
  });
});

app.post('/api/accounts', (req, res) => {
  const account = {
    id: accounts.length + 1,
    userId: req.user.sub,
    name: req.body.name,
    description: req.body.description,
    currencyId: req.body.currencyId,
    amount: req.body.amount,
    created: format(new Date()),
    updated: format(new Date())
  };
  accounts.push(account)
  res.json(account);
});

app.delete('/api/accounts/:id', (req, res) => {
  const account = accounts
    .filter((account) => account.userId === req.user.sub)
    .filter((account) => account.id === req.params.id)
    [0];
  if (account) {
    const index = accounts.indexOf(account);
    accounts.splice(index, 1);
  }
  res
    .status(204)
    .json();
});

app.get('/api/transactions', (req, res) => {
  res.json({
    transactions: transactions
      .filter(transaction => transaction.userId === req.user.sub)
  });
});

app.post('/api/transactions', (req, res) => {
  const account = {
    id: accounts.length + 1,
    userId: req.user.sub,
    fromAccountId: req.body.fromAccountId,
    toAccountId: req.body.toAccountId,
    rate: req.body.rate,
    name: req.body.name,
    description: req.body.description,
    fromAmount: req.body.fromAmount,
    toAmount: req.body.toAmount,
    categoryId: req.body.categoryId,
    date: req.body.date,
    created: format(new Date()),
    updated: format(new Date())
  };
  accounts.push(account)
  res.json(account);
});

app.delete('/api/transactions/:id', (req, res) => {
  const account = accounts
    .filter((account) => account.userId === req.user.sub)
    .filter((account) => account.id === req.params.id)
    [0];
  if (account) {
    const index = accounts.indexOf(account);
    accounts.splice(index, 1);
  }
  res
    .status(204)
    .json();
});

app.listen(app.get('port'), () => {
  console.log(`Find the server at: http://localhost:${app.get('port')}/`); // eslint-disable-line no-console
});