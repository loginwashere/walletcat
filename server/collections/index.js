const format = require('date-fns/format');

module.exports = () => {
  let items = [];
  const collection = {
    all: () => Promise.resolve(items),
    filterAll: cb => Promise.resolve(items.filter(cb)),
    allWithoutDeleted: () => collection.filterAll(i => !i.deleted),
    filter: cb => Promise.resolve(
      items
        .filter(i => typeof i.deleted === 'undefined')
        .filter(cb)
    ),
    find: id => collection.filter(u => u.id === id),
    filterOne: cb => collection
      .filter(cb)
      .then(items => Promise.resolve(items[0])),
    findOne: id => collection
      .find(id)
      .then(items => Promise.resolve(items[0])),
    add: (item) => {
      items = [
        ...items,
        item
      ];
      return Promise.resolve(item);
    },
    addUnique: (item, cb) => collection.filterAll(cb)
      .then(foundItem => {
        if (!foundItem.length) {
          return collection.add(item);
        }
        if (typeof foundItem[0].deleted === 'undefined') {
          return foundItem[0];
        }
        const newItem = Object.assign({}, item, { id: foundItem[0].id});
        items = [
          ...items.slice(0, items.indexOf(foundItem[0])),
          ...[newItem],
          ...items.slice(items.indexOf(foundItem[0]) + 1)
        ];
        return newItem;
      }),
    filterUpdate: (cb, params) => collection.filterOne(cb)
      .then(item => {
        if (!item) {
          return item;
        }
        const updatedItem = Object.assign({}, item, params);
        items[items.indexOf(item)] = updatedItem;
        return updatedItem;
      }),
    update: (id, params) => collection.filterUpdate(u => u.id === id, params),
    delete: id => {
      items = items.filter(i => i.id !== id);
      return Promise.resolve();
    },
    softDelete: id => collection.update(id, { deleted: format(new Date()) }),
    clean: () => {
      items = [];
      return Promise.resolve(items);
    }
  }
  return collection;
};
