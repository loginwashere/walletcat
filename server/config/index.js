//require('dotenv').config();

module.exports = {
  PORT: process.env.NODE_APP_PORT || process.env.PORT || 3001,
  HASH_SECRET: process.env.HASH_SECRET || 'abcdefg',
  JWT_SECRET: process.env.JWT_SECRET || 'shhhhhhared-secret',
  JWT_EXPIRES: process.env.JWT_EXPIRES || '1h',
  NODE_APP_SOCKET: process.env.NODE_APP_SOCKET
};