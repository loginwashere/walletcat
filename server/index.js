if (process.env.NODE_ENV === 'production') {
  require('newrelic')
}

const fs = require('fs')
const config = require('./config')
const server = require('./server')

const app = server()

app.set('port', config.PORT)
const httpServer = app.listen(config.NODE_APP_SOCKET || app.get('port'), () => {
  console.log(`Find the server at: http://localhost:${app.get('port')}/`) // eslint-disable-line no-console
  if (config.NODE_APP_SOCKET) {
    fs.openSync('/tmp/app-initialized', 'w')
  }
})

module.exports = {
  httpServer,
  app
}
