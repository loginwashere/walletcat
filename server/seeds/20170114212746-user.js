const format = require('date-fns/format')
const v4 = require('uuid/v4')
const hashPassword = require('../utils').hashPassword
const generateAvatarUrl = require('../utils').generateAvatarUrl

const users = [
  {
    id: v4(),
    email: 'admin@mail.com',
    username: 'admin',
    avatar: generateAvatarUrl('admin@mail.com'),
    password: hashPassword('qwerty'),
    createdAt: format(new Date()),
    updatedAt: format(new Date()),
  }
]

module.exports = {
  up: function (queryInterface, Sequelize) {
    return queryInterface.bulkInsert('users', users, {})
  },

  down: function (queryInterface, Sequelize) {
    return queryInterface.bulkDelete('users', null, {})
  },

  items: users
}
