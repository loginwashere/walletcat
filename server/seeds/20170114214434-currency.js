const format = require('date-fns/format')
const v4 = require('uuid/v4')

const currencies = [
  {
    id: v4(),
    name: 'USD',
    description: 'United States Dollar',
    createdAt: format(new Date()),
    updatedAt: format(new Date())
  },
  {
    id: v4(),
    name: 'UAH',
    description: 'Ukrainian Hryvna',
    createdAt: format(new Date()),
    updatedAt: format(new Date())
  }
]

module.exports = {
  up: function (queryInterface, Sequelize) {
    return queryInterface.bulkInsert('currencies', currencies, {})
  },

  down: function (queryInterface, Sequelize) {
    return queryInterface.bulkInsert('currencies', null, {})
  },

  items: currencies
};
