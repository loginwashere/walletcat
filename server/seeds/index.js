const format = require('date-fns/format');
const v4= require('uuid/v4');
const hashPassword = require('../utils').hashPassword;
const generateAvatarUrl = require('../utils').generateAvatarUrl;

const users = require('../collections/users');
const currencies = require('../collections/currencies');
const userCurrencies = require('../collections/userCurrencies');
const categories = require('../collections/categories');
const accounts = require('../collections/accounts');
const transactions = require('../collections/transactions');
const tokens = require('../collections/tokens');

const currenciesSeed = [
  {
    id: v4(),
    name: 'USD',
    description: 'United States Dollar',
    created: format(new Date()),
    updated: format(new Date())
  },
  {
    id: v4(),
    name: 'UAH',
    description: 'Ukrainian Hryvna',
    created: format(new Date()),
    updated: format(new Date())
  }
];

const usersSeed = [
  {
    id: v4(),
    email: 'admin@mail.com',
    username: 'admin',
    avatar: generateAvatarUrl('admin@mail.com'),
    password: hashPassword('qwerty'),
    created: format(new Date()),
    updated: format(new Date())
  }
];

const userCurrenciesSeed = [
  {
    id: v4(),
    // userId: users[0].id,
    // currencyId: currencies[0].id,
    created: format(new Date()),
    updated: format(new Date())
  }
];

const categoriesSeed = [
  {
    id: v4(),
    // userId: users[0].id,
    name: 'Fast food',
    description: 'Some fast food',
    created: format(new Date()),
    updated: format(new Date())
  }
];

const accountsSeed = [
  {
    id: v4(),
    // userId: users[0].id,
    // currencyId: userCurrencies[0].id,
    name: 'Wallet',
    description: 'My Wallet',
    amount: 0,
    created: format(new Date()),
    updated: format(new Date())
  },
  {
    id: v4(),
    // userId: users[0].id,
    // currencyId: userCurrencies[0].id,
    name: 'Stash',
    description: 'My Stash',
    amount: 0,
    created: format(new Date()),
    updated: format(new Date())
  }
];

const transactionsSeed = [
  {
    id: v4(),
    // userId: users[0].id,
    // fromAccountId: accounts[0].id,
    // toAccountId: accounts[1].id,
    // categoryId: categories[0].id,
    fromAmount: 100,
    toAmount: 100,
    description: 'description',
    date: format(new Date()),
    created: format(new Date()),
    updated: format(new Date())
  }
];

const applySeeds = () => {
  const cleamCollections = [
    currencies,
    users,
    userCurrencies,
    categories,
    accounts,
    transactions,
    tokens
  ].map(col => col['clean']());
  return Promise.all(cleamCollections)
    .then(result => Promise.all(currenciesSeed.map(currencies.add)))
    .then(currencies => {
      return Promise.all(usersSeed.map(users.add))
        .then(users => Promise.resolve(Object.assign({},{currencies}, {users})));
    })
    .then(result => {
      const { users, currencies } = result;
      const promises = userCurrenciesSeed
        .map(item => Object.assign({}, item, {
          userId: users[0].id,
          currencyId: currencies[0].id,
        }))
        .map(userCurrencies.add);
      return Promise.all(promises)
        .then(userCurrencies => Promise.resolve(Object.assign({}, result, {
          userCurrencies
        })));
    })
    .then(result => {
      const { users } = result;
      const promises = categoriesSeed
        .map(item => Object.assign({}, item, {
          userId: users[0].id,
        }))
        .map(categories.add);
      return Promise.all(promises)
        .then(categories => Promise.resolve(Object.assign({}, result, {
          categories
        })));
    })
    .then(result => {
      const { users, userCurrencies } = result;
      const promises = accountsSeed
        .map(item => Object.assign({}, item, {
          userId: users[0].id,
          currencyId: userCurrencies[0].id,
        }))
        .map(accounts.add);
      return Promise.all(promises)
        .then(accounts => Promise.resolve(Object.assign({}, result, {
          accounts
        })));
    })
    .then(result => {
      const { users, accounts, categories } = result;
      const promises = transactionsSeed
        .map(item => Object.assign({}, item, {
          userId: users[0].id,
          fromAccountId: accounts[0].id,
          toAccountId: accounts[1].id,
          categoryId: categories[0].id,
        }))
        .map(transactions.add);
      return Promise.all(promises)
        .then(transactions => Promise.resolve(Object.assign({}, result, {
          transactions
        })));
    })
    .then(result => {
      return Promise.resolve(result);
    });
};
module.exports = applySeeds;
