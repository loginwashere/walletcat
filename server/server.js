const path = require('path')
const express = require('express')
const bodyParser = require('body-parser')
const expressJwt = require('express-jwt')
const unless = require('express-unless')
const debug = require('debug')('app')
const config = require('./config')
const ExtandableError = require('./errors/extendable')
const ServerError = require('./errors/server-error')

module.exports = () => {
  const app = express()

  app.use((req, res, next) => {
    debug("req.header('origin')", req.header('origin'))
    res.setHeader('Access-Control-Allow-Headers', 'accept, authorization, content-type, x-requested-with')
    res.setHeader('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
    res.setHeader('Access-Control-Allow-Origin', '*')
    next()
  })

  const jwtMiddleware = expressJwt({
    secret: config.JWT_SECRET
  })
  .unless({
    path: [
      {
        url: /.*/,
        methods: ['OPTIONS']
      },
      {
        url: '/api/auth',
        methods: ['POST']
      },
      {
        url: '/api/users',
        methods: ['POST']
      },
      {
        url: '/api/users/email-confirm',
        methods: ['POST']
      },
      {
        url: '/api/users/resend-email-confirm',
        methods: ['POST']
      },
      {
        url: '/api/oauth/callback/facebook',
        methods: ['GET']
      }
    ]
  })

  app.use(bodyParser.json())
  app.use('/api/*', jwtMiddleware)

  // Express only serves static assets in production
  if (process.env.NODE_ENV === 'production') {
    const staticMiddleware = express.static('build')
    staticMiddleware.unless = unless
    app.use(staticMiddleware.unless({ method: 'OPTIONS' }))
  }

  const accountsRouter = require('./routes/accounts')
  const authRouter = require('./routes/auth')
  const currenciesRouter = require('./routes/currencies')
  const userCurrenciesRouter = require('./routes/userCurrencies')
  const categoriesRouter = require('./routes/categories')
  const transactionsRouter = require('./routes/transactions')
  const usersRouter = require('./routes/users')
  const oauthRouter = require('./routes/oauth')

  app.use('/api/accounts', accountsRouter)
  app.use('/api/auth', authRouter)
  app.use('/api/currencies', currenciesRouter)
  app.use('/api/user-currencies', userCurrenciesRouter)
  app.use('/api/categories', categoriesRouter)
  app.use('/api/transactions', transactionsRouter)
  app.use('/api/users', usersRouter)
  app.use('/api/oauth', oauthRouter)

  function removeStack(err) {
    const error = Object.assign({}, err)
    delete error.stack
    return error
  }

  app.use((err, req, res, next) => {
    debug(err)
    res.status(err.status || 500)
    const error = (app.get('env') === 'production')
      ? removeStack(err)
      : err
    if (error instanceof ExtandableError) {
      res.json(error)
    } else {
      res.json(new ServerError(error))
    }
    next(req)
  })

  const pathToIndex = path.resolve(`${__dirname}/../build/index.html`)
  app.get('*', (req, res) => {
    console.log(pathToIndex)
    res
      .status(200)
      .sendFile(pathToIndex)
  })
  return app
}
