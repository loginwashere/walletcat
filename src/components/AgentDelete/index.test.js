import React from 'react'
import renderer from 'react-test-renderer'
import { AgentDelete } from '.'
import { agentSeeder } from '../../../server/seeds'

describe('components:AgentDelete:', () => {
  const dispatch = jest.fn()
  const agentId = agentSeeder.items[0].id
  it('renders without crashing', () => {
    const tree = renderer.create(
      <AgentDelete dispatch={dispatch}
                   agentId={agentId} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    expect(dispatch.mock.calls.length).toBe(1)
  })
})
