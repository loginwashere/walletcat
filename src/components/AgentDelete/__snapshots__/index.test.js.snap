exports[`components:AgentDelete: renders without crashing 1`] = `
<div>
  <h1>
    Delete Category
  </h1>
</div>
`;
